### ISSUE TITLE
Add a flag to force running system calls

### ISSUE THREAD

**Issue by radu-gheorghe** [elastic/support-diagnostics#436]
This is a spin-off from https://github.com/elastic/support-diagnostics/issues/294#issuecomment-653078159

It looks like on some systems, support-diagnostics doesn't detect the bound address of ES, so it doesn't run system calls. I can reproduce this on an Ubuntu 18 Vagrant box, but I've seen it in many places. I think a flag that forces running system calls (maybe requiring the OS?) would be useful.

I'd gladly open a PR, just please let me know if there's a "right" way of doing this.

**Comment by nemonster** [comment]
Did you read the issue comments? I thought I explained what the process was and how it functions but perhaps it wasn't entirely clear. Three log lines isn't much to go on, and there's no information about the specific environment other than it's a Ubuntu 18.04 Vagrant box, which doesn't provide much background. Nor does "many places". OK, I guess..., but it also works in many places, so the issue would be more along the lines of _why_ doesn't run in those many places you're describing.

So my first question would be on the details of the cluster you're running this against and how you're running the utility to begin with? Platform? Container involved? One node cluster or multiples? Multiple hosts or one? Basic test mode or production configuration? If you've got some safe, non-threatening test output feel free to attach. If you can reproduce this as easily as you say, then I would think it's easy to either provide instructions to reproduce and/or output with a detailed description on how to set up the test bed.

The utility has to know what node it's on. For better or for worse, log collection is somewhat tied to the system calls and for that you need to be able to get the log location, which is in the output of the nodes call. So I check to make sure that the host the utility is running on actually matches up with the node that was input. Because you can use localhost, which is pretty amorphous, you could easily be running from one host and pointing at another. Which means that the system calls obtained are at best questionable and at worst misleading.
 
So I check the addresses that the JVM can find and compare them to the addresses bound to the node that's specified in the call. If I don't get a match (as explained above) it usually means you're pointing at a remote node or that there's some issue with the configuration where you don't get a match. 

Another question would be whether it works when you run the utility using the IP address? Or a hostname? Do you get them if you run with the remote option? See the docs for how to do that. If those work, that's a key to why you see this.

> Would it be useful to have a flag to force running system calls (update: added a separate issue for this part)? I'd love to submit such a PR if you give me a few pointers (I tried already once to patch DiagnosticChainExec and it didn't work).

I am never going to approve a PR to do this. Sorry, but this is used to diagnose cases by our engineers and there are valid reasons to prevent invalid output from showing up if possible. It muddies the water, so to speak and chews up time. There are actually too many options in this thing already IMO - and from the questions I get and the issues I see, I can say with absolute certainty that it is unusual for most people to spend much time with the documentation. So providing one more thing to create future issues is not high on my priority list. What you're asking for also has the potential to really screw up the remote call option. 

I have been considering adding a separate call that will run _only_ the system calls. An explicit, one operation command that just does for the sys calls what is happening for the REST api. Bundling a bunch of things so you don't have to type them all it. But that's as far as I want to take it.

See, but here's the thing. If you're that focused on this particular piece of functionality I think you're missing the big picture. Really. These calls run on one host. The one you're on, unless you're using the remote type. That's not even necessarily the host/node combo that has issues. And frankly, allowing someone to force that makes them even less useful, since you don't know for sure if it's even what you think it is. 

If I were still doing support cases I would be using the monitoring extract way more than this thing. At least to get started on a problem. If you want that kind of info, looking at the Metricbeat data in Kibana is going to provide a lot more clues to where a larger issue is than looking at the output from one host.  Because that data is time series and. because it's for _all the nodes_ not just one. If you see something where you need more detail you can then drill down with the host/call that's appropriate. 

If you would like to provide me more detail about what's going on I'm happy to look into this further. But this works for a lot of people and forcing it to run just because you have an issue without knowing why does not strike me as an appropriate course of action at least at this stage.

**Comment by radu-gheorghe** [comment]
Hello and thanks for the quick reply @nemonster!

First of all, sorry for being to brief here. I realize I came across as "not doing my homework at all", when in fact I was trying to be short :) Also, thanks a lot for your work on this useful tool - another implicit thing that shouldn't have been implicit. I'm very much appreciating it!

On to more details, here's one way to reproduce this issue: once you have Vagrant installed, you can do `vagrant init generic/ubuntu2004` to pull up the image of Ubuntu 20.04. I did this here just to use the latest versions. This will create a `Vagrantfile`, which I normally edit to add more memory and let it work over the bridged network (some of these steps are unnecessary, but it's close to how I had the "dirty" 18.04 box set up before):

```
$ grep -v '^\ *#' Vagrantfile | grep -v ^$
Vagrant.configure("2") do |config|
  config.vm.box = "generic/ubuntu2004"
  config.vm.network "public_network"
  config.vm.provider "virtualbox" do |vb|
     vb.memory = "4096"
  end
end
```

Then `vagrant up` will pull everything and start the VM. Use `vagrant ssh` to connect to it.

Once in, I set up ES and start it. Again, close to what I had before would be:
```
sudo apt-get update
sudo apt-get install openjdk-11-jdk
wget https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-oss-7.8.0-no-jdk-linux-x86_64.tar.gz
tar zxf elasticsearch-oss-7.8.0-no-jdk-linux-x86_64.tar.gz
cd elasticsearch-7.8.0/
mkdir -p jdk/bin
ln -s /usr/bin/java jdk/bin/java
bin/elasticsearch
```

Then download and run the diagnostics package:
```
wget https://github.com/elastic/support-diagnostics/releases/download/8.1.1/support-diagnostics-8.1.1-dist.zip
unzip support-diagnostics-8.1.1-dist.zip
cd support-diagnostics-8.1.1
```

At this point, if I run `sudo ./diagnostics.sh --host localhost`, system calls don't run. I'm attaching [diagnostics.log](https://github.com/elastic/support-diagnostics/files/4880286/diagnostics.log)

Before commenting here, I tried various things, such as making ES listen on 0.0.0.0 and using the "outside" IP address for the diagnostics tool, but none of these worked. However, if I install the same ES on my local Mac and run the diagnostics tool, it runs the system calls. As you say, I'm sure it works in lots of places (again, sorry for my inappropriate wording, I didn't mean to say that the tool is broken or another blanket statement).

Back to the issue at hand, what puzzles me is the empty list of `boundAddresses`. I assumed this is the key to the issue, but maybe I'm on the wrong track.

I realized that "forcing" the tool to run system calls may be a hack. It was just the only "easy" way I saw. And I offered to send a PR as a way to say "I'm not just expecting you to fix this". If there's any way I can help with this issue, please let me know.
